Khalyla requirement

petemeehan opened the pull request:
requirement:
1. given a sample set of pledges, choose the winning applicants. the total must be £2.5m or as close to that as possible, but not more
2. handle the case where some applicants are unable to pay by choosing new applicants from the original set. also this could happen multiple times in the future

petemeehan committed:
1st draft

petemeehan committed:
cleaned up a bit

cmck commented on monzo.js:
```diff
@@ -1,20 +1,83 @@
 
-var goal = 25000;
-var pledgeValues = [10,20,50,100,250,500,1000];
+var goal = 2500;
```
please change to 2,500,000 and retest

cmck commented on monzo.js:
```diff
@@ -1,20 +1,83 @@
-var goal = 25000;
-var pledgeValues = [10,20,50,100,250,500,1000];
+var goal = 2500;
+var pledgeValues = [10,20,50,100];
```
this has changed, why? people can pledge 10,20,50,100,250 or 1000

petemeehan committed:
changed goal to 2.5M

cmck commented on monzo.js:
```diff
@@ -1,20 +1,83 @@
-	datasetTotal -= randomlyChosenPledge;
+
+/*  Convert array to object so each pledge has an ID */
+dataset = dataset.reduce(function(obj, currentValue, index) {
```
it's not clear that this is test data

cmck commented on monzo.js:
```diff
@@ -1,20 +1,83 @@
 var pledgeValues = [10,20,50,100,250,500,1000];
 var maxPledgeValue = pledgeValues[pledgeValues.length - 1];
-var dataset = PD.sample(pledgeValues, 1000, true, [7,6,5,4,3,2,1]);
+var dataset = PD.sample(pledgeValues, 100000, true, [7,6,5,4,3,2,1]);
```
this line fails with a sample set size of 1000. The code should work regardless of the number of applicants

petemeehan replied:
it fails when not enough initial pledges to reach goal. this is a known bug. fix coming soon

petemeehan committed:
test commit - ignore

cmck commented on monzo.js:
```diff
@@ -1,20 +1,83 @@
-var dataset = PD.sample(pledgeValues, 1000, true, [7,6,5,4,3,2,1]);
+var dataset = PD.sample(pledgeValues, 1000, true, [7,6,5,4]);
 
 var datasetTotal = dataset.reduce((a, b) => a + b, 0);
```
ES6 only, you need a solution to this if the script is going to be used in a browser that doesn't support it

cmck commented on monzo.js:
```diff
@@ -1,20 +1,83 @@
+
+/*  Make copy of original dataset and total and randomly select pledges to
+    to remove so it now contains only chosen pledges */
+var chosenDatasetTotal = datasetTotal;
```
move algorithm to its own function

cmck commented on monzo.js:
```diff
@@ -1,20 +1,83 @@
+var chosenDatasetTotal = datasetTotal;
+var chosenDataset = {};
+for (var i in dataset) chosenDataset[i] = dataset[i]; // replace with Object.assign()?
+while (chosenDatasetTotal >= goal ) {
```
bug when goal=2.5m, dataset length 10k. you need to handle the case when less than 2.5m is raised

cmck replied:
also this section of code is slow try to work out why

cmck commented on monzo.js:
```diff
@@ -1,20 +1,83 @@
+var chosenDataset = {};
+for (var i in dataset) chosenDataset[i] = dataset[i]; // replace with Object.assign()?
+while (chosenDatasetTotal >= goal ) {
+  var ids = Object.keys(chosenDataset);
```
indenting

petemeehan replied:
elaborate?

cmck replied:
`var randomId = ids[Math.floor(Math.random() * ids.length)]` shouldn't be indented.

petemeehan replied:
apologies for not noticing this. fixed

cmck commented on monzo.js:
```diff
@@ -1,20 +1,83 @@
 }
+/*  Update original dataset by removing chosen applicants, so they can't
+    be chosen again later if more applicants are required */
+for (var i in chosenDataset) delete dataset[i];
```
move into its own function

cmck commented on monzo.js:
```diff
@@ -1,20 +1,83 @@
 
-console.log("Goal: " + goal);
+console.log("\nGoal: " + goal);
 console.log("Total pledged: " + datasetTotal);
```
look into the comma separation for string formatting in console.log()

cmck commented on monzo.js:
```diff
@@ -1,20 +1,83 @@
+
+
+/*  Remove failed pledges */
+var numberOfDroppedPledgers = 10;
```
numberOfDroppedPledgers needs to be 10% of the set of applications

cmck commented on monzo.js:
```diff
@@ -1,20 +1,83 @@
+
+
+
+function getRandomSubsetOfChosenApplicants(obj, numberOf) {
```
unclear what obj parameter is

cmck commented on monzo.js:
```diff
@@ -1,20 +1,83 @@
+  return subset;
+}
+
+function removeSubsetOfApplicants(obj, arrayOfApplicantIds) {
```
document these functions

petemeehan replied:
please advise whether new comments are sufficient

cmck commented:
test data:
```
var goal = 2500000;
var pledgeValues = [10,20,50,100,250,500,1000];
var maxPledgeValue = pledgeValues[pledgeValues.length - 1];
var dataset = PD.sample(pledgeValues, 30000, true, [7,6,5,4,3,2,1]);
```

`C:monzo-crowdfund clark$ node monzo.js
main algorithm Took **21635.5000** milliseconds

Goal: 2500000
Total pledged: 3804720
Number of chosen pledgers: 19442
Chosen pledges total: 2499860
getRandomSubsetOfChosenApplicants Took 1.5850 milliseconds
removeSubsetOfApplicants Took 0.1320 milliseconds

OH FUCK, WE LOST 10 PLEDGERS!
New number of chosen pledgers: 19432
New chosen pledges total: 2499370
repopulateChosenDataset Took 0.0820 milliseconds

NEW PLEDGERS SELECTED
New number of chosen pledgers: 19432
New chosen pledges total: 2499370`

petemeehan committed:
ignore

petemeehan committed:
Big commit - addressing all code reviews

1. Renamed many variables to be more descriptive

2. Added more comments, such as clarifying usage of test data

3. Removed ES6 features, ie. arrow functions

4. Amount of dropped pledgers is now 10% of the set of applications

5. Documented the functions

6. Fixed nested iterator bug, code now runs with 2.5M goal and 10k
initial pledges

petemeehan committed:
wrapped algorithms in functions

petemeehan committed:
added basic error handling

Throws warning string if insufficient pledges to reach goal

petemeehan reviewed (commented):
Need to catch exception if goal not reachable

petemeehan committed:
addressed PM's comments

1. No need for a comment describing the test data and other global vars
at the top, the code is self documenting in its simplicity. 2. Move
logging into the appropriate function, since it is related to it. 3.
`fractionOfDroppedPledgers` I think percentage is better than fraction
here. 4. Don't comment on function calls - the function itself has the
documentation.

cmck commented:
Lgtm
